raft/src/main/java/org/apache/kafka/raft/RequestManager.java

```diff
@@ -17,108 +17,296 @@
+     */
+    public long backoffBeforeAvailableBootstrapServer(long currentTimeMs) {
+        long minBackoffMs = retryBackoffMs;
+
+        Iterator<ConnectionState> iterator = connections.values().iterator();
+        while (iterator.hasNext()) {
+            ConnectionState connection = iterator.next();
+            if (connection.hasRequestTimedOut(currentTimeMs)) {
+                // Mark the node as ready after request timeout
+                iterator.remove();
+            } else if (connection.isBackoffComplete(currentTimeMs)) {
+                // Mark the node as ready after completed backoff
+                iterator.remove();
+            } else if (connection.hasInflightRequest(currentTimeMs)) {
+                // There can be at most one inflight fetch request
+                return connection.remainingRequestTimeMs(currentTimeMs);
             } else if (connection.isBackingOff(currentTimeMs)) {
                 minBackoffMs = Math.min(minBackoffMs, connection.remainingBackoffMs(currentTimeMs));
-            } else {
-                minBackoffMs = Math.min(minBackoffMs, connection.remainingRequestTimeMs(currentTimeMs));
             }
         }
+
+        // There are no inflight fetch requests so check if there is a ready bootstrap server
+        for (Node node : bootstrapServers) {
+            if (isReady(node, currentTimeMs)) {
+                return 0L;
+            }
+        }
+
+        // There are no ready bootstrap servers and inflight fetch requests, return the backoff
         return minBackoffMs;
     }
 
+    public boolean hasRequestTimedOut(Node node, long timeMs) {
+        ConnectionState state = connections.get(node.idString());
+        if (state == null) {
+            return false;
+        }
+
+        return state.hasRequestTimedOut(timeMs);
+    }
+
+    public boolean isReady(Node node, long timeMs) {
+        ConnectionState state = connections.get(node.idString());
+        if (state == null) {
+            return true;
+        }
+
+        boolean ready = state.isReady(timeMs);
+        if (ready) {
+            reset(node);
+        }
+
+        return ready;
+    }
+
+    public boolean isBackingOff(Node node, long timeMs) {
+        ConnectionState state = connections.get(node.idString());
+        if (state == null) {
+            return false;
+        }
+
+        return state.isBackingOff(timeMs);
+    }
+
+    public long remainingRequestTimeMs(Node node, long timeMs) {
+        ConnectionState state = connections.get(node.idString());
+        if (state == null) {
+            return 0;
+        }
+
+        return state.remainingRequestTimeMs(timeMs);
+    }
+
+    public long remainingBackoffMs(Node node, long timeMs) {
+        ConnectionState state = connections.get(node.idString());
+        if (state == null) {
+            return 0;
+        }
+
+        return  state.remainingBackoffMs(timeMs);
```

Comment:
nit: additional space after return.

Reply:
Thanks. Looks like I already removed this in another commit.